### PR TITLE
Iiif manifest

### PR DESCRIPTION
**Pull request by Dananji**

### CODE REVIEW

File: src/containers/StructureOutputContainer.js [hunk 1]
```diff
@@ -12,9 +12,11 @@ const StructureOutputContainer = (props) => {
...
   const apiUtils = new APIUtils();
 
-  const { structureURL, structureInfo, structuralMetadata } = props;
-  const { structureSaved } = structureInfo;
-  const { smData, initSmData, smDataIsValid } = structuralMetadata;
+  const dispatch = useDispatch();
+  const { manifestFetched } = useSelector((state) => state.manifest);
+  const { smData, initSmData, smDataIsValid } = useSelector((state) => state.structuralMetadata);
+  const { editingDisabled, structureInfo } = useSelector((state) => state.forms);
+  // const { smData, initSmData, smDataIsValid } = structuralMetadata;
```
Comment (cjcolvar) [inline src/containers/StructureOutputContainer.js:19]: Can this line be removed?

File: src/containers/StructureOutputContainer.js [hunk 2]
```diff
@@ -70,38 +69,38 @@ const StructureOutputContainer = (props) => {
...
+// });
 
-const mapDispatchToProps = {
-  postStructureSuccess: updateStructureStatus,
-  setAlert: setAlert,
-};
+// const mapDispatchToProps = {
+//   postStructureSuccess: updateStructureStatus,
+//   setAlert: setAlert,
+// };
```
Comment (cjcolvar) [inline src/containers/StructureOutputContainer.js:104]: Can these lines be removed?

File: src/services/WaveformDataUtils.js [hunk 3]
```diff
@@ -278,7 +272,7 @@ export default class WaveformDataUtils {
     // Check for margin of +/- 0.02 milliseconds to be considered
```
Comment (cjcolvar) [inline src/services/WaveformDataUtils.js:272]:
```suggestion
    // Check for margin of +/- 0.02 seconds to be considered
```

File: src/services/WaveformDataUtils.js [hunk 4]
```diff
@@ -416,7 +408,7 @@ export default class WaveformDataUtils {
...
```
Comment (cjcolvar) [inline src/services/WaveformDataUtils.js:408]:
```suggestion
    // Check for margin of +/- 0.02 seconds to be considered
```

File: src/services/iiif-parser.js [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+          children.push(structItem);
+        }
+      });
+    }
+  };
+
+  let manifestName;
+  let structures = [];
+
+  if (manifest != undefined || manifest != null) {
```
Comment (cjcolvar) [inline src/services/iiif-parser.js:130]: Should this be `&&`?
```suggestion
  if (manifest != undefined && manifest != null) {
```

File: src/services/iiif-parser.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+  else if (typeof initStructure === 'string' && initStructure !== '') {
+    structureJSON = smUtils.addUUIds([JSON.parse(initStructure)]);
+  }
+  // Use default initial structure when manifest doesn't
+  // have structures 
+  else if (initStructure != undefined &&
+    Object.keys(initStructure).length != 0) {
+    structureJSON = smUtils.addUUIds([initStructure]);
+  }
+  // Create a dummy structure with manifest information
```
Comment (cjcolvar) [inline src/services/iiif-parser.js:165]: How about rewording this?
```suggestion
  // Create an empty structure with manifest information
```

**Review by cjcolvar — COMMENTED** [review]
This looks great!

Is there any performance penalty with parsing the manifest multiple times in various methods in `iiif-parser.js`?  I'm wondering if it would make sense to parse it once and cache the result.

File: src/containers/StructureOutputContainer.js (hunk 1 above)
Comment (Dananji) [reply src/containers/StructureOutputContainer.js:19]: Yes :+1:

File: src/services/iiif-parser.js (hunk 5 above)
Comment (Dananji) [reply src/services/iiif-parser.js:130]: Absolutely! Good catch :100:
